pkg/util/limiter/rate_limiter.go

```diff
@@ -26,6 +26,17 @@ type RateLimiter struct {
 	tenants     map[string]*tenantLimiter
 }
 
+// Reservation is similar to rate.Reservation but excludes interfaces which do
+// not make sense to expose, because we are following the semantics of AllowN,
+// being an immediate reservation, i.e. not delayed into the future.
+type Reservation interface {
```

Comment:
Interface here is not necessary and is actually what adds extra allocation – which is what it tried to prevent in the first place. I would suggest to go back to previous design:

```go
type Reservation struct {
    r *Reservation
}
```

and return it directly as a value and not pointer (!). I would also suggest to add a unit test proving that we don't allocate more than necessary (see `testing.AllocsPerRun()`).

Reply:
That's unfortunate! Makes sense to change it back then.

Reply:
Turns out interface with pointer value doesn't need to do extra allocation, so my objection is not valid.

I still don't think we need the interface here... interfaces in Go are best used to give clients a way to provide different implementations. Interfaces don't hide the underlying implementation, as client code can still do `val.(*rate.Reservation)` on returned value. Custom struct as implemented originally achieves that goal better.

Reply:
Got confused about this so did a little experiment: https://play.golang.org/p/iHSGnvXzbEx

It seems assigning a pointer (to something which is already allocated) to an interface does not cause an additional allocation. This makes sense, as the interface is sized to hold pointers, but cannot be sized for structs of arbitrary size, hence in the struct case, a heap allocated copy of the struct must be made.